### PR TITLE
[Bugfix] UnicodeDecodeError: 'utf-8' codec can't decode byte

### LINKED ISSUE #113564: Torch Profiler with_stack=True, 'utf-8' codec can't decode
### 🐛 Describe the bug

I have been trying to use the pytorch profiler recently, under both the tensorboard profiler extensions analysis backend I received an error: UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf0 in position 38423127: invalid continuation byte. I have tried this multiple times and receive the same error, and under the Holistic trace analysis backend: UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf0 in position 38423161: invalid continuation byte. Think this could be related to: #64345, so I tried a nightly build which should in theory have the patch, although the issue persists.  

### Versions

Collecting environment information...
PyTorch version: 2.2.0.dev20231112+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.24.1
Libc version: glibc-2.35

Python version: 3.10.13 | packaged by conda-forge | (main, Oct 26 2023, 18:07:37) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-6.0.12-76060012-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA RTX A5000
GPU 1: NVIDIA GeForce RTX 4090

Nvidia driver version: 535.113.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   43 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          48
On-line CPU(s) list:             0-47
Vendor ID:                       AuthenticAMD
Model name:                      AMD Ryzen Threadripper 3960X 24-Core Processor
CPU family:                      23
Model:                           49
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       1
Stepping:                        0
Frequency boost:                 enabled
CPU max MHz:                     3800.0000
CPU min MHz:                     2200.0000
BogoMIPS:                        7585.73
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sev sev_es
Virtualization:                  AMD-V
L1d cache:                       768 KiB (24 instances)
L1i cache:                       768 KiB (24 instances)
L2 cache:                        12 MiB (24 instances)
L3 cache:                        128 MiB (8 instances)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-47
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] backpack-for-pytorch==1.6.0
[pip3] numpy==1.24.1
[pip3] pytorch-lightning==2.1.0
[pip3] pytorch-triton==2.1.0+6e4932cda8
[pip3] torch==2.2.0.dev20231112+cu118
[pip3] torch-tb-profiler==0.4.3
[pip3] torchaudio==2.2.0.dev20231112+cu118
[pip3] torchmetrics==1.2.0
[pip3] torchvision==0.17.0.dev20231112+cu118
[pip3] triton==2.1.0
[conda] backpack-for-pytorch      1.6.0                    pypi_0    pypi
[conda] blas                      2.16                        mkl    conda-forge
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libblas                   3.8.0                    16_mkl    conda-forge
[conda] libcblas                  3.8.0                    16_mkl    conda-forge
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch
[conda] liblapack                 3.8.0                    16_mkl    conda-forge
[conda] liblapacke                3.8.0                    16_mkl    conda-forge
[conda] mkl                       2020.2                      256  
[conda] numpy                     1.24.1                   pypi_0    pypi
[conda] pytorch-cuda              11.8                 h7e8668a_5    pytorch
[conda] pytorch-lightning         2.1.0                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] pytorch-triton            2.1.0+6e4932cda8          pypi_0    pypi
[conda] torch                     2.2.0.dev20231112+cu118          pypi_0    pypi
[conda] torch-tb-profiler         0.4.3                    pypi_0    pypi
[conda] torchaudio                2.2.0.dev20231112+cu118          pypi_0    pypi
[conda] torchmetrics              1.2.0                    pypi_0    pypi
[conda] torchtriton               2.1.0                     py310    pytorch
[conda] torchvision               0.17.0.dev20231112+cu118          pypi_0    pypi

### PR DESCRIPTION
**Pull request by Abatom**
Fixes #113564

When I used PyTorch's profiler to analyze the performance of vLLM, I encountered the following error. This error is similar to #113564. After analysis and troubleshooting, I changed the temporary file from text mode to binary mode, and it no longer reported an error and ran normally.

```bash
ERROR 10-28 10:25:50 engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 722, in stop 
ERROR 10-28 10:25:50 engine.py:160]     self._transit_action(self.current_action, None) 
ERROR 10-28 10:25:50 engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 751, in _transit_action 
ERROR 10-28 10:25:50 engine.py:160]     action() 
ERROR 10-28 10:25:50 engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 745, in _trace_ready 
ERROR 10-28 10:25:50 engine.py:160]     self.on_trace_ready(self) 
ERROR 10-28 10:25:50 engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 444, in handler_fn 
ERROR 10-28 10:25:50 engine.py:160]     prof.export_chrome_trace(os.path.join(dir_name, file_name)) 
ERROR 10-28 10:25:50 engine.py:160]   File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 220, in export_chrome_trace 
ERROR 10-28 10:25:50 engine.py:160]     fout.writelines(fin) 
ERROR 10-28 10:25:50 engine.py:160]   File "<frozen codecs>", line 322, in decode 
ERROR 10-28 10:25:50 engine.py:160] UnicodeDecodeError: 'utf-8' codec can't decode byte 0x8e in position 5896: invalid start byte
```

### CODE REVIEW

**Comment by Abatom** [comment]
> I'd like to accept this but I can't tell what the old code was trying to do

Thank you for your reply. What do I need to do?

**Comment by sraikund16** [comment]
> I'd like to accept this but I can't tell what the old code was trying to do

This function is called to create the JSON for perfetto/chrome tracer. If it sees that the user is trying to export to a .gz file, it creates a temporary file, exports the data to it using the autograd export_chrome_trace function, and then does compressed writes to the desired .gz file. 

I think in this case it makes the most sense to write in binary mode instead of text

File: torch/profiler/profiler.py [hunk 1]
```diff
@@ -242,11 +242,11 @@ def export_chrome_trace(self, path: str):
         """
         assert self.profiler
         if path.endswith(".gz"):
-            fp = tempfile.NamedTemporaryFile("w+t", suffix=".json", delete=False)
+            fp = tempfile.NamedTemporaryFile("w+b", suffix=".json", delete=False)
             fp.close()
             retvalue = self.profiler.export_chrome_trace(fp.name)
-            with open(fp.name) as fin:
-                with gzip.open(path, "wt") as fout:
+            with open(fp.name, 'rb') as fin:
```
Comment (sraikund16) [inline torch/profiler/profiler.py:248]: nit: can you use double quotes here since we use it for the surrounding lines?

Comment (Abatom) [reply torch/profiler/profiler.py]: > nit: can you use double quotes here since we use it for the surrounding lines?

Of course! Thank you for your reply.

**Review by ezyang — APPROVED** [review]

**Comment by tomasruizt** [comment]
thanks for fixing this bug! @Abatom. Do you know when it will be released?

**Comment by Abatom** [comment]
> thanks for fixing this bug! @Abatom. Do you know when it will be released?

sorry I don't know

**Comment by hchau630** [comment]
Also encountering this bug, will this fix be included in a release soon?

**Pull request closed without merging**